elevate-cpanel

```diff
@@ -2897,61 +2907,224 @@ sub _latest_checksum () {
+    my $grub2_info = $stage_info->{'grub2_workaround'};
+    if ( $grub2_info && $grub2_info->{'needs_workaround_update'} ) {
+        my $grub_dir  = GRUB2_PREFIX_DEBIAN;
+        my $grub2_dir = GRUB2_PREFIX_RHEL;
+
+        my $grub_bak;
+        if ( $grub2_info->{'backup_dir'} ) {
+            $grub_bak = $grub2_info->{'backup_dir'};
+        }
+        else {
+            $grub_bak = $grub2_info->{'backup_dir'} = $grub_dir . '-' . time;
+            update_stage_file( { 'grub2_workaround' => $grub2_info } );
+        }
+
+        rename $grub_dir, $grub_bak or LOGDIE("Unable to rename $grub_dir to $grub_bak: $!");    # failure on cross-device move is a feature
+        symlink $grub2_dir, $grub_dir or do {
+            rename $grub_bak, $grub_dir;                                                         # undo previous change on failure
+            LOGDIE("Unable to create symlink $grub_dir to point to $grub2_dir");                 # symlink() doesn't set $!
+        };
+
+    }
+
+    return;
+}
+
+sub merge_grub_directories_if_needed () {
+    my $stage_info = read_stage_file();
+    my $grub2_info = $stage_info->{'grub2_workaround'};
+    if ( $grub2_info && $grub2_info->{'needs_workaround_update'} ) {
+        my $grub_dir  = GRUB2_PREFIX_DEBIAN;
+        my $grub2_dir = GRUB2_PREFIX_RHEL;
+        my $grub_bak  = $grub2_info->{'backup_dir'};
+
+        # If something goes wrong here, we should give up on the copy but allow the upgrade to continue.
+        # Therefore, don't use LOGDIE() inside of the eval block.
+        eval {
+            my %grub2_entries;
+
+            opendir my $grub2_dir_fh, $grub2_dir or die "Unable to open directory $grub2_dir: $!";
+            while ( my $entry = readdir $grub2_dir_fh ) {
+                $grub2_entries{$entry} = 1;
+            }
+            closedir $grub2_dir_fh;
+
+            my ( $skipped_copy, $failed_copy ) = ( 0, 0 );
+
+            opendir my $grub_bak_fh, $grub_bak or die "Unable to open directory $grub_bak: $!";
+            while ( my $entry = readdir $grub_bak_fh ) {
+
+                next if $entry eq '.';
+                next if $entry eq '..';
+
+                # grub.cfg in the backup dir is the old symlink, so ignore it separately to avoid warnings:
+                next if $entry eq "grub.cfg";
+
+                if ( exists $grub2_entries{$entry} ) {
+                    $skipped_copy++;
+                    WARN("\"$grub_bak/$entry\" is not being copied to \"$grub2_dir/$entry\" because the destination already exists.");
+                    next;
+                }
+
+                if ( !File::Copy::Recursive::rcopy( "$grub_bak/$entry", "$grub2_dir/$entry" ) ) {
+                    $failed_copy++;
+                    WARN("Copying \"$grub_bak/$entry\" into \"$grub2_dir\" failed.");
+                }
+            }
+            closedir $grub_bak_fh;
+        };
+        WARN("Unable to copy the contents of \"$grub_bak\" into \"$grub2_dir\": $@") if $@;
```

Comment:
Does `File::Copy::Recursive::rcopy` die?

Reply:
It isn't documented to do so, nor do the functions it mentions it calls. I can verify this if wanted, of course.

Reply:
`opendir or die`